File: plugin/src/shared/sc/plugin2020/util/CubeCoordinates.java
package sc.plugin2020.util;

import org.jetbrains.annotations.NotNull;

public class CubeCoordinates implements Comparable {
  public final int x,y,z;

  public CubeCoordinates(int x, int y, int z){
    if (x + y + z != 0)
      throw new IllegalArgumentException("Constraint: (x + y + z == 0) not granted!");
    this.x = x;
    this.y = y;
    this.z = z;
  }

  public CubeCoordinates(int q, int r){
    this.x = q;
    this.y = r;
    this.z = -q-r;
  }

  public CubeCoordinates(CubeCoordinates position) {
    this.x = position.x;
    this.y = position.y;
    this.z = position.z;
  }

  @Override
  public String toString() {
    return String.format("(%d,%d,%d)", this.x, this.y, this.z);
  }

  @Override
  public int compareTo(@NotNull Object o) {
    CubeCoordinates to = (CubeCoordinates)o;
    if (this.x > to.x) return 1;
    if (this.x < to.x) return -1;
    if (this.y > to.y) return 1;
    if (this.y < to.y) return -1;
    if (this.z > to.z) return 1;
    if (this.z < to.z) return -1;
    return 0;
  }

  @Override
  public boolean equals(Object obj) {
    CubeCoordinates to = (CubeCoordinates)obj;
    return (this.x == to.x && this.y == to.y && this.z == to.z);
  }
}

File: plugin/src/shared/sc/plugin2020/util/CubeCoordinates.kt
package sc.plugin2020.util

import com.thoughtworks.xstream.annotations.XStreamAsAttribute
import kotlin.math.absoluteValue
import kotlin.math.sign

open class CubeCoordinates
@JvmOverloads constructor(
        @XStreamAsAttribute
        val x: Int,
        @XStreamAsAttribute
        val y: Int,
        @XStreamAsAttribute
        val z: Int = -x - y
): Comparable<CubeCoordinates> {
    
    constructor(position: CubeCoordinates): this(position.x, position.y, position.z)
    
    init {
        require(x + y + z == 0) { "Constraint: (x + y + z == 0) not fulfilled for ${this}!" }
    }
    
    override fun toString(): String = String.format("(%d,%d,%d)", x, y, z)
    
    override fun equals(other: Any?): Boolean =
            other is CubeCoordinates && x == other.x && y == other.y && z == other.z
    
    override fun hashCode(): Int =
            x * 31 + y
    
    fun distanceTo(other: CubeCoordinates) =  ((x - other.x).absoluteValue + (y - other.y).absoluteValue + (z - other.z).absoluteValue) / 2
    
    override operator fun compareTo(other: CubeCoordinates): Int {
        var sign = (x - other.x).sign
        if(sign == 0)
            sign = (y - other.y).sign
        return sign * distanceTo(other)
    }
    
}

